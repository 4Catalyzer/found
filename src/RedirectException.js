// @flow
// This isn't really an error.
export default class RedirectException {
  location: any;

  constructor(location: any) {
    this.location = location;
  }
}
