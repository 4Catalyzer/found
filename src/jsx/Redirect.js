// @flow
import invariant from 'invariant';

import RedirectObject from '../Redirect';

function createRoute(props) {
  return new RedirectObject(props);
}

function Redirect() {
  invariant(
    false,
    '<Redirect> is only for configuration and should not be rendered.',
  );
}

Redirect.createRoute = createRoute;

export default Redirect;
