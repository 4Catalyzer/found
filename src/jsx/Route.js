// @flow
import invariant from 'invariant';

function createRoute(props) {
  return props;
}

function Route() {
  invariant(
    false,
    '<Route> is only for configuration and should not be rendered.',
  );
}

Route.createRoute = createRoute;

export default Route;
