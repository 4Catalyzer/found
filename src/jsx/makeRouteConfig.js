import React from 'react';

export default function makeRouteConfig(node) {
  return React.Children.toArray(node)
    .filter(React.isValidElement)
    .map(({ type, props: { children, ...props } }) => {
      const route = type.createRoute(props);

      if (children) {
        route.children = makeRouteConfig(children);
      }

      return route;
    });
}
