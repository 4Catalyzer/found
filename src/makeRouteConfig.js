import React from 'react';

/**
 * Create a route configuration from JSX configuration elements.
 */
export default function makeRouteConfig(node) {
  return React.Children.toArray(node)
    .filter(React.isValidElement)
    .map(({ type: Type, props: { children, ...props } }) => {
      const route = new Type(props);

      if (children) {
        if (React.isValidElement(children) || Array.isArray(children)) {
          route.children = makeRouteConfig(children);
        } else {
          const routeGroups = {};
          Object.entries(children).forEach(([groupName, groupRoutes]) => {
            routeGroups[groupName] = makeRouteConfig(groupRoutes);
          });

          route.children = routeGroups;
        }
      }

      return route;
    });
}
