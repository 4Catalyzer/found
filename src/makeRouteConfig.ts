import React from 'react';

import { RouteConfig } from './typeUtils';

// TODO: what is even happening in this file? Try to fix as any's(+ maybe simplify)
function buildRouteConfig(
  node: React.ReactNode,
  routeConfig: any[],
): RouteConfig {
  React.Children.forEach(node, (child) => {
    // Falsy children get coerced to null. We check for this instead of
    // implicit falsiness because we don't want to allow empty strings or 0.
    if (child === null) {
      return;
    }

    if (!React.isValidElement(child)) {
      throw new TypeError(`\`${child}\` is not a valid React element`);
    }

    const Type = child.type as React.JSXElementConstructor<any>;
    const { children, ...props } = child.props;

    if (Type === React.Fragment) {
      buildRouteConfig(children, routeConfig);
      return;
    }

    const route = new (Type as any)(props);

    if (children) {
      if (React.isValidElement(children) || Array.isArray(children)) {
        route.children = buildRouteConfig(children, []);
      } else {
        const routeGroups: Record<string, RouteConfig> = {};
        Object.entries(children).forEach(([groupName, groupRoutes]: any[]) => {
          routeGroups[groupName] = buildRouteConfig(groupRoutes, []);
        });

        route.children = routeGroups;
      }
    }

    routeConfig.push(route);
  });

  return routeConfig;
}

/**
 * Create a route configuration from JSX configuration elements.
 */
export default function makeRouteConfig(node: React.ReactNode): RouteConfig {
  return buildRouteConfig(node, []);
}
