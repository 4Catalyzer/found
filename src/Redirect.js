import RedirectException from './RedirectException';

export default class Redirect {
  constructor({ from, to }) {
    this.path = from;
    this.to = to;
  }

  render({ match }) {
    const to = this.to;
    let toLocation;

    if (typeof to === 'function') {
      toLocation = to(match);
    } else {
      const { router, params } = match;
      toLocation = router.matcher.format(to, params);
    }

    throw new RedirectException(toLocation);
  }
}

if (__DEV__) {
  // Workaround to make React Proxy give me the original class, to allow
  // makeRouteConfig to get the actual class, when using JSX for routes.
  Redirect.prototype.isReactComponent = {};
}
