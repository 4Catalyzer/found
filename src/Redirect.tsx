/* eslint-disable max-classes-per-file */
import React from 'react';

import RedirectException from './RedirectException';
import { LocationDescriptor, Match, RedirectOptions } from './typeUtils';

class Redirect implements RedirectOptions {
  path?: string;

  to: string | ((match: Match) => LocationDescriptor);

  status?: number;

  constructor({ from, to, status }: RedirectOptions) {
    this.path = from;
    this.to = to;
    this.status = status;
  }

  render({ match }: { match: Match }) {
    const { to, status } = this as any;
    let toLocation;

    if (typeof to === 'function') {
      toLocation = to(match);
    } else {
      const { router, params } = match;
      toLocation = router.matcher.format(to, params);
    }

    throw new RedirectException(toLocation, status);
  }
}

if (__DEV__) {
  // Workaround to make React Proxy give me the original class, to allow
  // makeRouteConfig to get the actual class, when using JSX for routes.
  (Redirect.prototype as any).isReactComponent = {};
}

// This actually doesn't extend a React.Component, but we need consumer to think that it does
declare class RedirectType extends React.Component<RedirectOptions> {
  constructor(config: RedirectOptions);
}

export default Redirect as unknown as typeof RedirectType;
