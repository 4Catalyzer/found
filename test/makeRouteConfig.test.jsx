import React from 'react';
import { describe, expect, it } from 'vitest';

import Redirect from '../src/Redirect';
import RedirectException from '../src/RedirectException';
import Route from '../src/Route';
import makeRouteConfig from '../src/makeRouteConfig';

describe('makeRouteConfig', () => {
  const AppPage = () => {};

  const MainPage = () => {};
  const FooPage = () => {};
  const BarPage = () => {};
  const BazPage = () => {};

  const FooNav = () => {};
  const FooA = () => {};
  const FooB = () => {};
  const BarNav = () => {};
  const BarMain = () => {};

  it('should work with a route', () => {
    expect(makeRouteConfig(<Route path="/" Component={AppPage} />)).toEqual([
      new Route({
        path: '/',
        Component: AppPage,
      }),
    ]);
  });

  it('should work with nested routes', () => {
    expect(
      makeRouteConfig(
        <Route path="/" Component={AppPage}>
          <Route Component={MainPage} />
          {/* This comment should be ignored. */}
          <Route path="foo" Component={FooPage}>
            <Route path="bar" Component={BarPage} />
          </Route>
        </Route>,
      ),
    ).toEqual([
      new Route({
        path: '/',
        Component: AppPage,
        children: [
          new Route({
            Component: MainPage,
          }),
          new Route({
            path: 'foo',
            Component: FooPage,
            children: [
              new Route({
                path: 'bar',
                Component: BarPage,
              }),
            ],
          }),
        ],
      }),
    ]);
  });

  it('should work with fragments', () => {
    expect(
      makeRouteConfig(
        <Route path="/" Component={AppPage}>
          <Route Component={MainPage} />
          <>
            <Route path="foo" Component={FooPage} />
            <Route path="bar" Component={BarPage} />
          </>
          <Route path="baz" Component={BazPage} />
        </Route>,
      ),
    ).toEqual([
      new Route({
        path: '/',
        Component: AppPage,
        children: [
          new Route({
            Component: MainPage,
          }),
          new Route({
            path: 'foo',
            Component: FooPage,
          }),
          new Route({
            path: 'bar',
            Component: BarPage,
          }),
          new Route({
            path: 'baz',
            Component: BazPage,
          }),
        ],
      }),
    ]);
  });

  it('should work with <Redirect>', () => {
    expect(
      makeRouteConfig(
        <Route path="/" Component={AppPage}>
          <Redirect from="widget/:widgetId" to="/widgets/:widgetId" />
        </Route>,
      ),
    ).toEqual([
      new Route({
        path: '/',
        Component: AppPage,
        children: [
          new Redirect({
            from: 'widget/:widgetId',
            to: '/widgets/:widgetId',
          }),
        ],
      }),
    ]);
  });

  it('should work with named child routes', () => {
    expect(
      makeRouteConfig(
        <Route path="/" Component={AppPage}>
          <Route path="foo">
            {{
              nav: <Route path="(.*)?" Component={FooNav} />,
              main: [
                <Route path="a" Component={FooA} />,
                <Route path="b" Component={FooB} />,
              ],
            }}
          </Route>
          <Route path="bar">
            {{
              nav: <Route path="(.*)?" Component={BarNav} />,
              main: <Route Component={BarMain} />,
            }}
          </Route>
        </Route>,
      ),
    ).toEqual([
      {
        path: '/',
        Component: AppPage,
        children: [
          {
            path: 'foo',
            children: {
              nav: [
                {
                  path: '(.*)?',
                  Component: FooNav,
                },
              ],
              main: [
                {
                  path: 'a',
                  Component: FooA,
                },
                {
                  path: 'b',
                  Component: FooB,
                },
              ],
            },
          },
          {
            path: 'bar',
            children: {
              nav: [
                {
                  path: '(.*)?',
                  Component: BarNav,
                },
              ],
              main: [
                {
                  Component: BarMain,
                },
              ],
            },
          },
        ],
      },
    ]);
  });

  it('should error on non-element children', () => {
    expect(() => {
      makeRouteConfig(
        <Route path="/" Component={AppPage}>
          {/* The comma below will fail the invariant. */}
          <Route path="foo" Component={FooPage} />,
          <Route path="bar" Component={BarPage} />
        </Route>,
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('should allow empty children', () => {
    expect(() => {
      makeRouteConfig(
        <Route path="/" Component={AppPage}>
          {false}
          <Route path="foo" Component={FooPage} />
        </Route>,
      );
    }).not.toThrow();
  });

  it('should error on other falsy children', () => {
    expect(() => {
      makeRouteConfig(
        <Route path="/" Component={AppPage}>
          {0}
          <Route path="foo" Component={FooPage} />
        </Route>,
      );
    }).toThrowErrorMatchingSnapshot();
  });

  ['react-proxy', 'react-stand-in'].forEach((packageName) => {
    it(`should work with proxies from ${packageName}`, () => {
      // eslint-disable-next-line global-require, import/no-dynamic-require
      const createProxy = require(packageName).default;

      const ProxiedRedirect = createProxy(Redirect).get();
      const redirect = makeRouteConfig(
        <ProxiedRedirect from="/foo" to="/bar" />,
      )[0];

      expect(Object.getPrototypeOf(redirect)).toBe(Redirect.prototype);

      expect(redirect.path).toBe('/foo');
      expect(redirect.to).toBe('/bar');
      expect(redirect.render).toBeTruthy();

      let redirectException;

      try {
        redirect.render({
          match: {
            router: {
              matcher: {
                format: (to) => to,
              },
            },
          },
        });
      } catch (error) {
        redirectException = error;
      }

      expect(redirectException).toBeInstanceOf(RedirectException);
      // expect(redirectException.url).toBe('/bar');
    });
  });
});
