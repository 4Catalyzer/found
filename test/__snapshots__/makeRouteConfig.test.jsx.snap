// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`makeRouteConfig > should error on non-element children 1`] = `[TypeError: \`,\` is not a valid React element]`;

exports[`makeRouteConfig > should error on other falsy children 1`] = `[TypeError: \`0\` is not a valid React element]`;
